### PR TITLE
B-21581: Incorrect GBLOC on Move When Origin Duty Location is Updated

### PR DESCRIPTION
**Pull request by WeatherfordAaron**
## Testing Steps I took
Use [this doc](https://dp3.atlassian.net/wiki/download/attachments/2279735297/postal_code_to_gblocs_202311091157.csv?api=v2) to determine what zip codes to use in order to see different GBLOCs.
1. As a service member (to include CIV or other), create a move with a shipment. Pay special attention to what selection is made for your origin duty location.
2. As SC and/or TOO verify which GBLOC shows up in the Origin GBLOC column for the move when you search for it. Also verify that when you open the move as an office user, the GBLOC should also show in the top-right of the page as you are looking at the move/shipment.
3. As either the service member or as an office user, change the origin duty location to that which falls under a different GBLOC (reference the document link above which is from the helpful files section of the ByteSize M&Ms Confluence pages).
4. As SC and/or TOO verify that the GBLOC has changed.

### CODE REVIEW

**Review by brooklyn-welsh — CHANGES_REQUESTED** [review]
Tried to test by first creating a KKFA move, creating a shipment, and then switching origin location to Bloomingdale, IL 60108 (BGNC).

While the GBLOC did change for the move, the move is *not* routed to the BGNC counselor queue. It *is* however removed from the KKFA queue (or rather, just doesn't show up there in the first place).

Please test and confirm, but from my testing this needs to be changed to fill the 2nd part of the AC. 

P.S. Tested routing by creating a move that was BGNC with same origin from the beginning, and that move was routed to the BGNC queue correctly, which points to some logic missing in the GBLOC change code somewhere.

**Comment by WeatherfordAaron** [comment]
> Tried to test by first creating a KKFA move, creating a shipment, and then switching origin location to Bloomingdale, IL 60108 (BGNC).
> 
> While the GBLOC did change for the move, the move is _not_ routed to the BGNC counselor queue. It _is_ however removed from the KKFA queue (or rather, just doesn't show up there in the first place).
> 
> Please test and confirm, but from my testing this needs to be changed to fill the 2nd part of the AC.
> 
> P.S. Tested routing by creating a move that was BGNC with same origin from the beginning, and that move was routed to the BGNC queue correctly, which points to some logic missing in the GBLOC change code somewhere.

I believe I figured this part out and updated the code so that it should route correctly upon updating as the customer or as the counselor

File: pkg/handlers/ghcapi/internal/payloads/model_to_payload.go [hunk 1]
```diff
@@ -2454,19 +2454,15 @@ func SearchMoves(appCtx appcontext.AppContext, moves models.Moves) *ghcmessages.
...
+		if gblocErr != nil {
+			originGBLOC = *ghcmessages.NewGBLOC("")
+		} else if customer.Affiliation.String() == "MARINES" {
+			originGBLOC = ghcmessages.GBLOC("USMC/" + OriginPostalCodeToGBLOC.GBLOC)
 		} else {
-			// If the move's first shipment doesn't have a pickup address (like with an NTS-Release),
-			// we need to fall back to the origin duty location GBLOC.  If that's not available for
-			// some reason, then we should get the empty string (no GBLOC).
-			originGBLOC = swag.StringValue(move.Orders.OriginDutyLocationGBLOC)
+			originGBLOC = ghcmessages.GBLOC(OriginPostalCodeToGBLOC.GBLOC)
```
[the remark this answers is not in the document]
Comment (WeatherfordAaron) [reply pkg/handlers/ghcapi/internal/payloads/model_to_payload.go:2465]: let me find it but there's a separate method that handles that already (IIRC).. I think someone just copied that to this when it doesn't work right which is why this ticket got created in the first place

File: pkg/handlers/internalapi/orders.go [hunk 2]
```diff
@@ -364,6 +364,15 @@ func (h UpdateOrdersHandler) Handle(params ordersop.UpdateOrdersParams) middlewa
...
 
+				originGBLOC, originGBLOCerr := models.FetchGBLOCForPostalCode(appCtx.DB(), originDutyLocation.Address.PostalCode)
+				if originGBLOCerr != nil {
+					return handlers.ResponseForError(appCtx.Logger(), originGBLOCerr), originGBLOCerr
+				}
+				order.OriginDutyLocationGBLOC = &originGBLOC.GBLOC
+				if originGBLOCerr != nil {
+					return handlers.ResponseForError(appCtx.Logger(), originGBLOCerr), originGBLOCerr
+				}
+
```
Comment (taeJungCaci) [inline pkg/handlers/internalapi/orders.go:375]: As discussed - reverting changes in model_to_payload and adding tests for this would do the trick I think.

Comment (WeatherfordAaron) [reply pkg/handlers/internalapi/orders.go:375]: I'm going to add a playwright test since the test coverage isn't tripping

Comment (taeJungCaci) [reply pkg/handlers/internalapi/orders.go:375]: Shouldn't it be the backend test to cover the code you've added - and to help avoid the coverage issue in the future?

Comment (WeatherfordAaron) [reply pkg/handlers/internalapi/orders.go:375]: Yea I may need to add that too.. that to protect the code change.. but the playwright test would be more to protect the actual ask from the PO

**Review by JamesHawks224 — APPROVED** [review]
testing: success
code: lgtm

File: pkg/models/move.go [hunk 3]
```diff
@@ -413,6 +413,49 @@ func SaveMoveDependencies(db *pop.Connection, move *Move) (*validate.Errors, err
...
+
+	var originGBLOC PostalCodeToGBLOC
+	originGBLOC, err = FetchGBLOCForPostalCode(db, originPostalCode)
+	if err != nil {
+		return "", err
+	}
+
+	return originGBLOC.GBLOC, nil
+}
+
```
Comment (taeJungCaci) [inline pkg/models/move.go:458]: Not sure where these are getting used at.

File: pkg/handlers/internalapi/orders_test.go [hunk 4]
```diff
@@ -584,143 +584,247 @@ func (suite *HandlerSuite) TestUploadAmendedOrdersHandlerIntegration() {
...
+			LinkOnly: true,
+		}}, nil)
 
-	suite.Run("Can update CONUS and OCONUS orders", func() {
-		testCases := []struct {
-			isOconus bool
-		}{
-			{isOconus: true},
-			{isOconus: false},
-		}
```
Comment (taeJungCaci) [inline pkg/handlers/internalapi/orders_test.go:594]: Not sure if you meant to delete this test.

**Review by taeJungCaci — CHANGES_REQUESTED** [review]
Also I think the server_test might be failing and possibly not flaky

File: pkg/handlers/internalapi/orders_test.go (hunk 4 above)
Comment (WeatherfordAaron) [reply pkg/handlers/internalapi/orders_test.go]: I did not intend this. I fixed it

**Comment by WeatherfordAaron** [comment]
> Tried to test by first creating a KKFA move, creating a shipment, and then switching origin location to Bloomingdale, IL 60108 (BGNC).
> 
> While the GBLOC did change for the move, the move is _not_ routed to the BGNC counselor queue. It _is_ however removed from the KKFA queue (or rather, just doesn't show up there in the first place).
> 
> Please test and confirm, but from my testing this needs to be changed to fill the 2nd part of the AC.
> 
> P.S. Tested routing by creating a move that was BGNC with same origin from the beginning, and that move was routed to the BGNC queue correctly, which points to some logic missing in the GBLOC change code somewhere.

This should be resolved now

File: pkg/models/move.go (hunk 3 above)
Comment (WeatherfordAaron) [reply pkg/models/move.go:458]: you are correct. Removing this

Comment (WeatherfordAaron) [reply pkg/models/move.go]: removed them

**Review by taeJungCaci — CHANGES_REQUESTED** [review]

File: pkg/factory/duty_location_factory.go [hunk 5]
```diff
@@ -124,6 +124,51 @@ func BuildDutyLocationWithoutTransportationOffice(db *pop.Connection, customs []
...
+					PostalCode:     "35023",
+				},
+				Type: &Addresses.DutyLocationAddress,
+			},
+		}, nil)
+	}
+
+	return defaultLocation
+}
+
```
[the remark this answers is not in the document]
Comment (WeatherfordAaron) [reply pkg/factory/duty_location_factory.go:171]: you are correct.. removed

File: pkg/handlers/internalapi/orders.go [hunk 6]
```diff
@@ -354,16 +354,36 @@ func (h UpdateOrdersHandler) Handle(params ordersop.UpdateOrdersParams) middlewa
...
+			if originGBLOCerr != nil {
+				return handlers.ResponseForError(appCtx.Logger(), originGBLOCerr), originGBLOCerr
+			}
+
+			order.OriginDutyLocationGBLOC = &originGBLOC.GBLOC
+			if originGBLOCerr != nil {
+				return handlers.ResponseForError(appCtx.Logger(), originGBLOCerr), originGBLOCerr
+			}
+
+			newDutyLocationGBLOC, err := models.FetchGBLOCForPostalCode(appCtx.DB(), destinationDutyLocation.Address.PostalCode)
```
Comment (taeJungCaci) [inline pkg/handlers/internalapi/orders.go:386]: There seems to be a lot of duplicate code here that are already in `if payload.OriginDutyLocationID != "" {` case after line 393 - which are code for `originDutyLocation`. You can reuse code in there to fetch `originDutyLocation`. 

So bottom line - you can achieve the same result by adding the following code under line 401:

```
originGBLOC, originGBLOCerr := models.FetchGBLOCForPostalCode(appCtx.DB(), originDutyLocation.Address.PostalCode)
if originGBLOCerr != nil {
    return handlers.ResponseForError(appCtx.Logger(), originGBLOCerr), originGBLOCerr
}
order.OriginDutyLocationGBLOC = &originGBLOC.GBLOC
```

File: pkg/handlers/internalapi/orders_test.go [hunk 7]
```diff
@@ -723,6 +723,130 @@ func (suite *HandlerSuite) TestUpdateOrdersHandler() {
...
+
+	suite.NoError(okResponse.Payload.Validate(strfmt.Default))
+	suite.Equal(string(newOrdersType), string(*okResponse.Payload.OrdersType))
+
+	fetchedOrder, err = models.FetchOrder(suite.DB(), order.ID)
+	suite.NoError(err)
+
+	fetchedPostalCode, err = fetchedOrder.GetOriginPostalCode(suite.DB())
+	suite.NoError(err)
+	fetchedGBLOC, err = fetchedOrder.GetOriginGBLOC(suite.DB())
```
Comment (taeJungCaci) [inline pkg/handlers/internalapi/orders_test.go:843]: Also this should be `fetchedGBLOC = *fetchedOrder.OriginDutyLocationGBLOC` to check if order gbloc was changed. You are currently fetching new Gbloc.

**Review by taeJungCaci — CHANGES_REQUESTED** [review]

File: pkg/models/order.go [hunk 8]
```diff
@@ -309,6 +309,49 @@ func (o *Order) CreateNewMove(db *pop.Connection, moveOptions MoveOptions) (*Mov
...
+
+	var originGBLOC PostalCodeToGBLOC
+	originGBLOC, err = FetchGBLOCForPostalCode(db, originPostalCode)
+	if err != nil {
+		return "", err
+	}
+
+	return originGBLOC.GBLOC, nil
+}
+
```
Comment (danieljordan-caci) [inline pkg/models/order.go:354]: Add tests please

**Review by danieljordan-caci — CHANGES_REQUESTED** [review]

File: pkg/models/order.go (hunk 8 above)
Comment (danieljordan-caci) [reply pkg/models/order.go:354]: Gotcha. No worries!

File: pkg/models/order.go [hunk 9]
```diff
@@ -309,6 +309,49 @@ func (o *Order) CreateNewMove(db *pop.Connection, moveOptions MoveOptions) (*Mov
 	return createNewMove(db, *o, moveOptions)
 }
 
+/*
+ * GetOriginPostalCode returns the GBLOC for the postal code of the the origin duty location of the order.
+ */
+func (o Order) GetOriginPostalCode(db *pop.Connection) (string, error) {
+	// Since this requires looking up the order in the DB, the order must have an ID. This means, the order has to have been created first.
+	if uuid.UUID.IsNil(o.ID) {
+		return "", errors.WithMessage(ErrInvalidOrderID, "You must created the order in the DB before getting the origin GBLOC.")
```
Comment (paulstonebraker) [inline pkg/models/order.go:318]: Typo here? I think you meant 'create' vs 'created' here and below on 338.

**Review by paulstonebraker — CHANGES_REQUESTED** [review]
Just a small typo and you're good to go!

**Review by paulstonebraker — APPROVED** [review]

**Pull request merged**